opnode/rollup/driver/state.go

```diff
@@ -2,118 +2,200 @@ package driver
 
 import (
 	"context"
+	"time"
 
 	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
 	"github.com/ethereum-optimism/optimistic-specs/opnode/rollup"
 	"github.com/ethereum/go-ethereum/log"
 )
 
-// internalDriver exposes the driver functionality that maintains the external execution-engine.
-type internalDriver interface {
-	// requestEngineHead retrieves the L2 he```d reference of the engine, as well as the L1 reference it was derived from.
-	// An error is returned when the L2 head information could not be retrieved (timeout or connection issue)
-	requestEngineHead(ctx context.Context) (refL1 eth.BlockID, refL2 eth.BlockID, err error)
-	// findSyncStart statelessly finds the next L1 block to derive, and on which L2 block it applies.
-	// If the engine is fully synced, then the last derived L1 block, and parent L2 block, is repeated.
-	// An error is returned if the sync starting point could not be determined (due to timeouts, wrong-chain, etc.)
-	findSyncStart(ctx context.Context) (nextRefL1 eth.BlockID, refL2 eth.BlockID, err error)
-	// driverStep explicitly calls the engine to derive a L1 block into a L2 block, and apply it on top of the given L2 block.
-	// The finalized L2 block is provided to update the engine with finality, but does not affect the derivation step itself.
-	// The resulting L2 block ID is returned, or an error if the derivation fails.
-	driverStep(ctx context.Context, nextRefL1 eth.BlockID, refL2 eth.BlockID, finalized eth.BlockID) (l2ID eth.BlockID, err error)
+type inputInterface interface {
+	L1Head(ctx context.Context) (eth.L1BlockRef, error)
+	L2Head(ctx context.Context) (eth.L2BlockRef, error)
+	L1ChainWindow(ctx context.Context, base eth.BlockID) ([]eth.BlockID, error)
+	// SafeL2Head is the L2 Head found via the sync algorithm
+	SafeL2Head(ctx context.Context) (eth.L2BlockRef, error)
 }
 
-type state struct {
-	// l1Head tracks the L1 block corresponding to the l2Head
-	l1Head eth.BlockID
+type outputInterface interface {
+	step(ctx context.Context, l2Head eth.BlockID, l2Finalized eth.BlockID, l1Window []eth.BlockID) (eth.BlockID, error)
+}
 
-	// l2Head tracks the head-block of the engine
-	l2Head eth.BlockID
+type state struct {
+	// Chain State
+	l1Head      eth.BlockID   // Latest recorded head of the L1 Chain
+	l1Base      eth.BlockID   // L1 Parent of L2 Head block
```

Comment:
So I renamed `L1Parent` -> `L1origin`. I don't love `L1Origin` for this because it is not any generic `L1Origin`, it is the `L1Origin` of the `L2SafeHead`. Maybe origin is a little better, but I would not name this `L1Parent`